operator/pkg/resources/headless_service.go

```diff
@@ -166,3 +166,84 @@ func (r *HeadlessServiceResource) getAnnotation() map[string]string {
 		externalDNSUseHostIP: "true",
 	}
 }
+
+// RenderHeadlessService creates a headless Service for the Redpanda cluster
+func RenderHeadlessService(cluster *vectorizedv1alpha1.Cluster) *corev1.Service {
+	headlessPorts := generateHeadlessServicePorts(cluster)
+
+	ports := make([]corev1.ServicePort, 0, len(headlessPorts))
+	for _, svcPort := range headlessPorts {
+		ports = append(ports, corev1.ServicePort{
+			Name:       svcPort.Name,
+			Protocol:   corev1.ProtocolTCP,
+			Port:       int32(svcPort.Port),
+			TargetPort: intstr.FromInt32(int32(svcPort.Port)),
+		})
+	}
+
+	objLabels := labels.ForCluster(cluster)
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        cluster.Name,
+			Namespace:   cluster.Namespace,
+			Labels:      objLabels,
+			Annotations: getHeadlessServiceAnnotations(cluster),
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: "v1",
+		},
+		Spec: corev1.ServiceSpec{
+			PublishNotReadyAddresses: true,
+			Type:                     corev1.ServiceTypeClusterIP,
+			ClusterIP:                corev1.ClusterIPNone,
+			Ports:                    ports,
+			Selector:                 objLabels.AsAPISelector().MatchLabels,
+		},
+	}
+}
+
+// generateHeadlessServicePorts collects all internal ports for the headless service
+func generateHeadlessServicePorts(cluster *vectorizedv1alpha1.Cluster) []NamedServicePort {
+	var headlessPorts []NamedServicePort
+
+	adminAPI := cluster.AdminAPIInternal()
+	if adminAPI != nil {
+		headlessPorts = append(headlessPorts, NamedServicePort{
+			Name: AdminPortName,
+			Port: adminAPI.Port,
+		})
+	}
```

Comment:
nit: It's preferable to use inline assignments and checks here. Doing so indicates to the reader that the variable will only be used within this if block and allows you to use more terse naming.

```suggestion
	if port := cluster.AdminAPIInternal(); port != nil {
		headlessPorts = append(headlessPorts, NamedServicePort{
			Name: AdminPortName,
			Port: port.Port,
		})
	}
```